src/Connection/CountLimitingConnectionPool.php

```diff
@@ -0,0 +1,306 @@
+                $this->removeWaiting($uri, $deferredId);
+            });
+
+            if ($this->shouldMakeNewConnection($uri)) {
+                break;
+            }
+
+            $connection = yield $deferredPromise;
+
+            \assert($connection instanceof Connection);
+
+            $stream = yield $this->getStreamFrom($connection, $request);
+
+            if ($stream === null) {
+                continue; // Wait for a different connection to become available.
+            }
+
+            return [$connection, $stream];
+        } while (true);
+
+        $this->totalConnectionAttempts++;
+
+        $connectionPromise = $this->connectionFactory->create($request, $cancellation);
+
+        $connectionId = \spl_object_id($connectionPromise);
+        $this->connections[$uri] = $this->connections[$uri] ?? new \ArrayObject;
+        $this->connections[$uri][$connectionId] = $connectionPromise;
+
+        $connectionPromise->onResolve(function (?\Throwable $exception, ?Connection $connection) use (
+            &$deferred,
+            $uri,
+            $connectionId,
+            $isHttps
+        ): void {
+            if ($exception) {
+                $this->dropConnection($uri, $connectionId);
+                if ($deferred !== null) {
+                    $deferred->fail($exception); // Fail Deferred so Promise\first() below fails.
+                }
+                return;
+            }
+
+            $this->openConnectionCount++;
+
+            if ($isHttps) {
+                $this->waitForPriorConnection[$uri] = \in_array('2', $connection->getProtocolVersions(), true);
+            }
+
+            $connection->onClose(function () use ($uri, $connectionId): void {
+                $this->openConnectionCount--;
+                $this->dropConnection($uri, $connectionId);
+            });
+        });
+
+        try {
+            $connection = yield Promise\first([$connectionPromise, $deferredPromise]);
+        } catch (MultiReasonException $exception) {
+            [$exception] = $exception->getReasons(); // The first reason is why the connection failed.
+            throw $exception;
+        }
+
+        $deferred = null; // Null reference so connection promise handler does not double-resolve the Deferred.
+        $this->removeWaiting($uri, $deferredId); // Deferred no longer needed for this request.
+
+        \assert($connection instanceof Connection);
+
+        $stream = yield $this->getStreamFrom($connection, $request);
+
+        if ($stream === null) {
+            // Reused connection did not have an available stream for the given request.
+            $connection = yield $connectionPromise; // Wait for new connection request instead.
+
+            $stream = yield $this->getStreamFrom($connection, $request);
+
+            // New connection should always return a stream for the request.
+            \assert($stream instanceof Stream);
```

Comment:
This might fail, because other requests can subscribe to the same connection promise and we're no longer the first one to subscribe here, so another waiting requests might be "quicker" (executed first) and take the remaining streams.

Reply:
Yep, that could happen I suppose. Fixed.